### ISSUE TITLE
`react/prop-types` false positive on arrow function definition

### ISSUE THREAD

**Issue by SevenOutman** [yannickcr/eslint-plugin-react#2971]
![image](https://user-images.githubusercontent.com/8225666/115396942-1dee3f80-a218-11eb-984f-7edd4692a5eb.png)

`react/prop-types` recognizes an arrow function definition as function component.

The line shown in the screenshot is https://github.com/rsuite/rsuite/blob/faf62a0594a483ced9fc795a54ace34573b16c10/src/TreePicker/TreePicker.tsx#L717

Run into this issue with version `7.23.2`. Downgraded to `7.20.3` and it's gone.

**Comment by golopot** [comment]
Some thoughts:

1. The warning goes away if you give it prop types, something like `({key, index, style}: {key: string, index: number, style: any})`.
2. We need to decide whether or not this rule should warn about components defined in a higher order component.

**Comment by ljharb** [comment]
I think it should indeed; an HOC returns a component, and that component needs propTypes.

**Comment by Akiyamka** [comment]
As I understand same issue in my code
![image](https://user-images.githubusercontent.com/6587585/127889916-b82ad031-f127-44a0-aba9-e6c0e1476c3b.png)

```ts
interface HighlightedLinkProps {
  to: string;
  label: string;
  exact?: boolean;
}

function HighlightedLink({ to, label, exact }: HighlightedLinkProps) {
  const match = useRouteMatch({
    path: to,
    exact,
  });

  return (
    <div className={match ? style.activeLink : ''}>
      <Link to={to}>{label}</Link>
    </div>
  );
}

export default function NavigationMenu({ links }: { links: HighlightedLinkProps[] }) {
  return (
    <nav className={style.navigation}>
      {links.map((props) => (
        <HighlightedLink key={props.to} {...props} />
      ))}
    </nav>
  );
}

```

**Comment by aurimasniekis** [comment]
I am getting the same error when I am not even doing propTypes, I have a render function that receives object type via generic and returns ReactNode and it complains about props validation.

```ts
import React from 'react';

interface SomeType<ContextType = any> {
  renderValue: (context: ContextType) => React.ReactNode;
}

interface DataObject {
  id: string,
  title: string,
  value: string,
}

const someType: SomeType<DataObject> = {
  renderValue: ({title}) => <div>{title}</div>,
};
```

it goes away if I add type to the argument:

```ts
const someType: SomeType<DataObject> = {
  renderValue: ({title}: DataObject) => <div>{title}</div>,
};
```

But this way whole generic approach and type hinting loses purpose...
